### ISSUE TITLE
Treat signed overflows more cautiously in branching

### ISSUE THREAD

**Issue by jerhard** [goblint/analyzer#163]
### Background

For the implementation of integer domains, we decided to treat the operations that might result in signed overflows as resulting in undefined values.

### Problem

If we want to be really strict in treating signed overflows as resulting in undefined values, there are some nasty implications for our treatment of branching. These concern the  refinement we do with `invariant` and `inv_bin_int`. 

Strictly speaking, if the evaluation of an expression might have led to a signed overflow, we then cannot use the expression used for branching to refine the values.

##### Example:

```C
#include <assert.h>
int main(){
    int x;
    // We cannot exclude that x+1 results in a signed overflow
    // Thus, we cannot conclude that x+1 == 2 implies that x == 1
    if(x + 1 == 2){
        assert(x==1); // UNKNOWN!
    }
}
```
In the example, we will currently conclude that the assertion holds (and is not unknown). It is not quite clear, whether and in what sense this is sound. 

### Suggested Solution
By default, we don't refine values on branching expressions when we cannot exclude a signed overflow. The current behavior could be activated by an option.

### Remark
Signed overflows actually result in undefined behavior in the C standard (not only undefined values). So what we should do here (and whether we should change our implementation at all) is really not so clear.

**Comment by vogler** [comment]
Accessing `x` (uninitialized) is already undefined behavior.
Is there an analysis to find undefined behavior or an option to go to top in these cases?
IMO the default configuration should behave as gcc which should be what we're currently doing, i.e. uninitialized access gives top and `x` is refined to 1.
In the example there's no possible other value for `x` even in case of overflow.
